### ISSUE TITLE
Gossiping of warrants to agent_activity authorities

### ISSUE THREAD

**Issue by abe-njama** [holochain/holochain#3576]
In the effort to prevent spam, we have the problem of having the agent_activity authorities know that someone has done something.

Current Problem:
- Agents autonomously decide on warranting, leading to inconsistencies in identifying bad behaviour.
- Lack of coordination between agents results in gaps where some agents may not know about warranted entries.
- Requestors of specific data may not receive information about bad behaviour due to lack of coordination.

Proposed Solution:
- Bob signs a warrant for a problematic entry and shares it with Carol, Alice's agent authority.
- Carol validates the entry upon receiving its header, blocking and warranting Alice if deemed necessary.
- Carol gossips the warrant to other agent activity authorities.
- If Carol disagrees with the warrant, she warrants Bob and forwards the warrant to Edward, Bob's agent activity authority.
- Edward independently validates the warrant and takes appropriate actions.
- Dave, requesting Alice's agent activity from Carol, will now receive warrant information about Alice's bad entry, resolving gaps in information dissemination.
- The proposed solution eliminates potential DOS vectors and ensures consistency in identifying and addressing bad behaviour.

**Comment by maackle** [comment]
In conversation with @artbrock we decided that the best way to implement Warrants is to make Warrant a new type of Op. Then in the scenario above, when Bob shares a warrant with Carol, that is simply Bob publishing the Warrant to its proper place (the basis hash of the Alice's agent key, which coincides with Alice's Agent authority). And when Carol gossips with other agent authorities, that is simply the same thing as normal Op gossip.

So this issue main boils down to adding the new Op type(s) for Warrants. We get the propagation of said Warrants for free.

**Comment by maackle** [comment]
On the topic of adding Warrant as a new op type...I think as we move into this kind of thing, with one authority authoring an Op intended specifically for another authority, the three basic authority types -- Record, Entry, and Agent Activity (Chain) -- come more clearly into focus, and it becomes more important that they line up. Rather than add ever more op types, I propose we restructure DhtOp something like this:

(I also propose we rename Agent Activity Authority to Chain Authority, just because it's shorter, but I'm at least going to use that for this code snippet)

```rust
// Pull the common signature out of the Op enum
pub struct Op {
    op: AuthorityOp,
    signature: Signature,
}

// Split Op up into the three authority types.
// Then list all the ops that that authority can handle.
enum AuthorityOp {
    RecordAuthority(RecordAuthorityOp)
    EntryAuthority(EntryAuthorityOp)
    ChainAuthority(ChainAuthorityOp)
}

// Ops come in two basic flavors:
// - Data: an Author wants the authority to store some data on their chain
// - Warrant: an Authority is signaling to another Authority that some data is invalid

enum RecordAuthorityOp {
    Data(RecordAuthorityData),
    Warrant(RecordAuthorityWarrant),
}

enum EntryAuthorityOp {
    Data(EntryAuthorityData),
    Warrant(EntryAuthorityWarrant),
}

enum ChainAuthorityOp {
    Data(ChainAuthorityData),
    Warrant(ChainAuthorityWarrant),
}

// The data types that we already have can be rearranged a bit to fit the new
// authority-centric structure

enum RecordAuthorityData {
    /// StoreRecord
    Store(Action, RecordEntry),

    /// RegisterUpdatedRecord, i.e. update the Action
    Update(action::Update, RecordEntry),

    /// RegisterAddLink, i.e. add a link whose base is an Action
    AddLink(action::CreateLink),

    /// RegisterRemoveLink
    RemoveLink(action::DeleteLink),
}

enum EntryAuthorityData {
    /// StoreEntry
    Store(NewEntryAction, Entry),

    /// RegisterUpdatedContent, i.e. update the Entry
    Update(action::Update, RecordEntry),

    /// RegisterAddLink, i.e. add a link whose base is an Entry
    AddLink(action::CreateLink),

    /// RegisterRemoveLink
    RemoveLink(action::DeleteLink),
}

enum ChainAuthorityOp {
    /// RegisterAgentActivity
    Store(Action),

    Warrant(ChainAuthorityWarrant),
}

// We expect to have a variety of Warrant types issued at the system level,
// this gives room for expansion.

enum RecordAuthorityWarrant {}

enum EntryAuthorityWarrant {}

enum ChainAuthorityWarrant {
    /// Proof of chain fork
    ChainFork(SignedAction, SignedAction)
}
```

Then we have a clear distinction of which ops are "data" and which are "warrants", and we also know the proper groupings. A Warrant can only have an effect on the serving of data under the same authority. For instance, the output of `get_agent_activity` is only be affected by the collection of all `ChainAuthorityOp`s collected by an authority, other ops do not factor into this function's output. Same for all other HDK functions asking for data from the DHT, the outputs are purely determined by the set of ops that fall under one of the 3 authority types.

**Comment by maackle** [comment]
In terms of writing actual acceptance criteria for this, we should first come to agreement on (1) if we want to add Warrant to Op and (2) how exactly we want to do that. I believe once we do that, then everything else is already done.

**Comment by thedavidmeister** [comment]
@ThetaSinner if you're talking about sybil resistance, that's generally a happ design level concern like membrane proofs etc.

**Comment by ThetaSinner** [comment]
- Warrants as ops, need to be added
- Basis for warrant ops uses the agent so gossip should get warrants to the right place without changes

**Comment by maackle** [comment]
latest AC:

- add new Warrant variant to Op (https://hackmd.io/@hololtd/BJk0R1Dx0?type=view)
- add ability to `get` DhtOpHashes (at all) and `get` ActionHashes from agent authorities instead of record authorities
- on receiving a warrant that you find to be valid, block the "subject" of the warrant (the author who made the bad data)
- re-publish invalid warrants with warrant author as target (counter-warranting)

follow-on:

- transitive warranting in app-specific cases (detecting agents who have been knowingly interacting with warranted agents)

**Comment by maackle** [comment]
Closed by #3875
